web/packages/teleport/src/JoinTokens/CreateJoinTokenDialog.tsx

```diff
@@ -0,0 +1,565 @@
+          service_accounts: rule.service_accounts.map(account => account.value),
+        })),
+      };
+      request.gcp = gcp;
+    }
+
+    runCreateTokenAttempt(request);
+  }
+
+  function setTokenRoles(roles: Option<JoinRole, JoinRole>[]) {
+    setNewTokenState(prevState => ({
+      ...prevState,
+      roles,
+    }));
+  }
+
+  function setTokenMethod(method: Option<JoinMethod, JoinMethod>) {
+    // set the method and reset the token rules per type for a fresh form
+    setNewTokenState(prevState => ({
+      ...prevState,
+      method,
+      iam: [{ aws_account: '', aws_arn: '' }], // default
+    }));
+  }
+
+  function setTokenField(fieldName: string, value: string) {
+    setNewTokenState(prevState => ({
+      ...prevState,
+      [fieldName]: value,
+    }));
+  }
+
+  return (
+    <Flex width="500px">
+      <Box
+        pr={4}
+        pl={4}
+        css={`
+          overflow: auto;
+          width: 100%;
+          min-height: 50vh;
+          padding-bottom: 0;
+        `}
+      >
+        <Flex
+          alignItems="center"
+          mb={3}
+          justifyContent="space-between"
+          maxWidth={maxWidth}
+        >
+          <Flex alignItems="center" mr={3}>
+            <HoverTooltip tipContent="Back to Join Tokens">
+              <ButtonIcon onClick={onClose} mr={2} ml={'-8px'}>
+                <Cross size="medium" />
+              </ButtonIcon>
+            </HoverTooltip>
+            <Text typography="h3" fontWeight={400}>
+              {editToken ? `Edit Token` : 'Create a New Join Token'}
+            </Text>
+          </Flex>
+          {editToken && (
+            <ButtonText
+              p={2}
+              onClick={() => {
+                onClose();
+                editTokenWithYAML(editToken.id);
+              }}
+            >
+              <Text color="buttons.link.default">Use YAML editor instead</Text>
+            </ButtonText>
+          )}
+        </Flex>
+        <Validation>
+          {({ validator }) => (
+            <Box maxWidth={maxWidth}>
+              {createTokenAttempt.status === 'error' && (
+                <Alert kind="danger">{createTokenAttempt.statusText}</Alert>
+              )}
+              <FieldSelect
+                label="Method"
+                rule={requiredField('Please select a join method')}
+                isSearchable
+                isSimpleValue
+                isClearable={false}
+                value={newTokenState.method}
+                onChange={setTokenMethod}
+                options={availableJoinMethods}
+              />
+              {newTokenState.method.value !== 'token' && ( // if the method is token, we generate the name for them on the backend
+                <FieldInput
+                  label="Token name"
+                  labelTip={
+                    editToken
+                      ? 'Editing the token name will create a new token and not remove the current one.'
+                      : ''
```

Comment:
i think this is a surprising side affect, especially when using with yaml b/c it doesn't warn me. every where else with resource editing in UI, we consistently prevent user from re-naming an existing resource and backend returns an error like `resource naming not supported create another or something` and i think this is the best way to do it.

i think we should not use upsert, but instead use explicit create and edit, and on edit, disable editing name

Reply:
Ok I'll add some new endpoints. Thank you!